libcaf_core/caf/blocking_fan_out_response_handle.hpp

```
@@ -0,0 +1,346 @@
+      = [pending{helper.pending}, timeout{state_.in_flight},
+         g{std::forward<OnError>(on_error)}](error& err) mutable {
+          auto lg = log::core::trace("pending = {}", *pending);
+          if (*pending > 0) {
+            timeout.dispose();
+            *pending = 0;
+            g(err);
+          }
+        };
+    auto bhvr = behavior{std::move(helper), std::move(error_handler)};
+    for (auto mid : state_.mids) {
+      auto now = std::chrono::steady_clock::now();
+      auto remaining = state_.deadline > now
+                         ? (state_.deadline - now)
+                         : std::chrono::steady_clock::duration::zero();
+      state_.self->do_receive(mid, bhvr, remaining);
+    }
+  }
+
+  template <class OnValue, class OnError>
+  void receive(OnValue on_value, OnError on_error) &&
+    requires is_select_any
+  {
+    detail::fan_out_response_type_check<Policy, OnValue, OnError, Results...>();
+    using helper = caf::detail::select_any_factory<std::decay_t<OnValue>>;
+    auto pending = std::make_shared<size_t>(state_.mids.size());
+    auto result_handler = helper::make(pending, state_.in_flight,
+                                       std::forward<OnValue>(on_value));
+    auto error_handler
+      = [p{std::move(pending)}, timeout{state_.in_flight},
+         g{std::forward<OnError>(on_error)}](error& err) mutable {
+          if (*p == 0) {
+            // nop
+          } else if (*p == 1) {
+            timeout.dispose();
+            g(err);
+          } else {
+            --*p;
+          }
+        };
+    auto bhvr = behavior{std::move(result_handler), std::move(error_handler)};
+    for (const auto& mid : state_.mids) {
+      auto now = std::chrono::steady_clock::now();
+      auto remaining = state_.deadline > now
+                         ? (state_.deadline - now)
+                         : std::chrono::steady_clock::duration::zero();
+      state_.self->do_receive(mid, bhvr, remaining);
+    }
```

This loop kinda defeats the point of `select_any`, because it will wait until all responses arrive before returning. It should return immediately at the first (valid) response. Of course we can't just return here after the first match because that would "leak" the other responses. I think the only way to do this cleanly is to add a list of "to-be-ignored" message IDs to the blocking actor. Then it silently discards the response for that response once it arrives. Feels a bit hacky, but I couldn't come up with a better idea.

The `select_all` case also needs to return early if it receives an error. Same logic: mark the message IDs to be discarded.

---

Thanks for the review. Of course that is a problem. The reason is that I wanted to have the implementation not modify the `abstract_blocking_actor*` interface, but as you suggested above, this will have to change the state of the blocking actor inevitably, so I'll try refactoring it with the suggestions.

---

I have a [`in_flight` disposable ](https://github.com/actor-framework/actor-framework/pull/2132/changes#diff-9d15f7b81e3fb057df9a73a1517ed1589ca6ca4ef42b6c3f8e6df2ea90278c12R80)in the `state_`. It's a [composite disposable created with](https://github.com/actor-framework/actor-framework/pull/2132/changes#diff-3ad181645d1bfa7aeff44e516eef7a0697f8bb19bc31fdc2cdb85c1f8a907c53R143) all the disposables returned from `self->request_response_timeout` in one case, and `clock.schedule_message` in the other case. Could returning early and disposing all remaining messages via this disposable be a simpler way to fix this?